google-cloud-firestore/src/main/java/com/google/cloud/firestore/AggregateQuery.java

```diff
@@ -136,28 +213,45 @@ private final class AggregateQueryResponseObserver
 
     private final AggregateQueryResponseDeliverer responseDeliverer;
     private StreamController streamController;
+    private Timestamp readTime = Timestamp.MAX_VALUE;
+    @Nullable private Map<String, Value> aggregateFieldsMap = null;
+    @Nullable private ExplainMetrics metrics = null;
 
     AggregateQueryResponseObserver(AggregateQueryResponseDeliverer responseDeliverer) {
       this.responseDeliverer = responseDeliverer;
     }
 
+    private boolean isExplainQuery() {
+      return this.responseDeliverer.explainOptions != null;
+    }
+
     @Override
     public void onStart(StreamController streamController) {
       this.streamController = streamController;
     }
 
     @Override
     public void onResponse(RunAggregationQueryResponse response) {
-      // Close the stream to avoid it dangling, since we're not expecting any more responses.
-      streamController.cancel();
+      if (response.hasReadTime()) {
+        readTime = Timestamp.fromProto(response.getReadTime());
+      }
 
-      // Extract the aggregations and read time from the RunAggregationQueryResponse.
-      Timestamp readTime = Timestamp.fromProto(response.getReadTime());
+      if (response.hasResult()) {
+        aggregateFieldsMap = response.getResult().getAggregateFieldsMap();
+      }
+
+      if (response.hasExplainMetrics()) {
+        metrics = new ExplainMetrics(response.getExplainMetrics());
+      }
 
-      // Deliver the result; even though the `RunAggregationQuery` RPC is a "streaming" RPC, meaning
-      // that `onResponse()` can be called multiple times, it _should_ only be called once. But even
-      // if it is called more than once, `responseDeliverer` will drop superfluous results.
-      responseDeliverer.deliverResult(response.getResult().getAggregateFieldsMap(), readTime);
+      if (!isExplainQuery()) {
+        // Deliver the result; even though the `RunAggregationQuery` RPC is a "streaming" RPC,
+        // meaning that `onResponse()` can be called multiple times, it _should_ only be called
+        // once for non-explain queries. But even if it is called more than once,
+        // `responseDeliverer` will drop superfluous results. For explain queries, there will
+        // be more than one response, and the last response will contain the metrics.
+        onComplete();
+      }
```

Comment:
Can we just omit this and let the normal `onComplete` shine through? This deviation from convention doesn't add value as far as I can tell.

Reply:
I agree with point about optimization.